tm2/pkg/bft/privval/signer/remote/client/conn.go

```diff
@@ -0,0 +1,143 @@
+package client
+
+import (
+	"net"
+	"time"
+
+	"github.com/gnolang/gno/tm2/pkg/amino"
+	r "github.com/gnolang/gno/tm2/pkg/bft/privval/signer/remote"
+)
+
+// ensureConnection tries to establish a connection with the server.
+func (rsc *RemoteSignerClient) ensureConnection() error {
+	// If the connection is already established, return.
+	if rsc.isConnected() {
+		rsc.logger.Debug("Already connected to server")
+		return nil
+	}
+
+	// Try to establish a connection with the server, retrying if necessary.
+	for try := 0; ; try++ {
+		// Ensure the client is not closed.
+		if rsc.isClosed() {
+			return nil
+		}
+
+		// Dial the server.
+		conn, err := net.DialTimeout(rsc.protocol, rsc.address, rsc.dialTimeout)
+		if err != nil {
+			rsc.logger.Warn("Failed to dial",
+				"protocol", rsc.protocol,
+				"address", rsc.address,
+				"error", err,
+			)
+
+			// If the maximum retries are not exceeded, log attempts count then retry.
+			if rsc.dialMaxRetries > 0 && try < rsc.dialMaxRetries {
+				rsc.logger.Info("Retrying to connect", "try", try+1, "maxRetry", rsc.dialMaxRetries)
+			} else if rsc.dialMaxRetries < 0 { // Retry indefinitely.
+				rsc.logger.Info("Retrying to connect", "try", try+1, "maxRetry", "unlimited")
+			} else { // Max retries exceeded.
+				return ErrMaxRetriesExceeded
+			}
+
+			// Wait for the retry interval before trying again.
+			time.Sleep(rsc.dialRetryInterval)
+			continue
+		}
+		rsc.logger.Debug("Dial succeeded")
+
+		// If the connection is a TCP connection, configure and secure it.
+		tcpConn, ok := conn.(*net.TCPConn)
+		if ok {
+			// Configure and secure the TCP connection then authenticate the server.
+			sconn, err := r.ConfigureTCPConnection(
+				tcpConn,
+				rsc.clientPrivKey,
+				rsc.authorizedKeys,
+				rsc.keepAlivePeriod,
+				rsc.requestTimeout,
+			)
+			if err != nil {
+				rsc.logger.Error("Failed to configure TCP connection", "error", err)
+				conn.Close() // Close the connection if its configuration failed.
+				return err
+			}
+
+			rsc.logger.Debug("Configured TCP connection successfully")
+			conn = sconn
+		}
+
+		// Set the connection.
+		rsc.setConnection(conn)
+		rsc.logger.Info("Connected to server", "protocol", rsc.protocol, "address", rsc.address)
+
+		return nil
+	}
+}
+
+// isConnected returns true if the client is connected to the server.
+func (rsc *RemoteSignerClient) isConnected() bool {
+	rsc.connLock.RLock()
+	defer rsc.connLock.RUnlock()
+	return rsc.conn != nil
+}
+
+// setConnection sets the connection to the server.
+func (rsc *RemoteSignerClient) setConnection(conn net.Conn) error {
```

Comment:
This method feels suuuuuper sketchy

Reply:
Can you elaborate on why it's needed?

Reply:
What is sketchy about it? I'll try to summarize, but basically:
- This client implementation ensures on every request it sends that its connected to the server (otherwise it tries to reconnect)
- If the dialling to the server succeeded, it then store a reference to this connection, so:
  - The client knows its connection to the server is valid (there is no `⁠isConnected` method or anything similar in the `net.Conn` interface)
  - When the node needs to close the client, we need this ref to close the underlying conn

Reply:
Got it, so it's because we need an active reference to the connection (it changes)